### PR TITLE
chore: add authentication for ApisixRoute

### PR DESCRIPTION
**Pull request by tao12345666333**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues

___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

File: pkg/kube/apisix/apis/config/v2alpha1/types.go [hunk 1]
```diff
@@ -194,6 +195,16 @@ type ApisixRouteHTTPPlugin struct {
 // any plugins.
 type ApisixRouteHTTPPluginConfig map[string]interface{}
 
+type ApisixRouteAuthentication struct {
```
Comment (tokers) [inline pkg/kube/apisix/apis/config/v2alpha1/types.go:198]: Should add some comments to explain the usage of this type.

File: pkg/kube/apisix/apis/config/v2alpha1/types.go [hunk 2]
```diff
@@ -194,6 +195,16 @@ type ApisixRouteHTTPPlugin struct {
...
+	Enable  bool                             `json:"enable" yaml:"enable"`
+	Type    string                           `json:"type" yaml:"type"`
+	KeyAuth ApisixRouteAuthenticationKeyAuth `json:"keyauth,omitempty" yaml:"keyauth,omitempty"`
+}
+
+type ApisixRouteAuthenticationKeyAuth struct {
```
Comment (tokers) [inline pkg/kube/apisix/apis/config/v2alpha1/types.go:204]: Ditto.

File: test/e2e/features/consumer.go [hunk 3]
```diff
@@ -57,5 +54,146 @@ spec:
...
+
+		msg = s.NewAPISIXClient().GET("/ip").
+			WithHeader("Host", "httpbin.org").
+			WithHeader("X-Foo", "baz").
+			WithHeader("apikey", "baz").
+			Expect().
+			Status(http.StatusNotFound).
+			Body().
+			Raw()
+		assert.Contains(ginkgo.GinkgoT(), msg, "404 Route Not Found")
```
Comment (tokers) [inline test/e2e/features/consumer.go:197]: Should also add test cases for:

1. Use secret to fetch basic auth and key auth parameters.
2. Test with authentication `false`

Comment (tao12345666333) [reply test/e2e/features/consumer.go:293]: Addressed.

File: samples/deploy/crd/v1beta1/ApisixRoute.yaml [hunk 4]
```diff
@@ -225,7 +225,6 @@ spec:
                             type: string
                     required:
                       - enable
-                      - type
```
Comment (tokers) [inline samples/deploy/crd/v1beta1/ApisixRoute.yaml]: `type` should be a required field. Why change this?

Comment (tao12345666333) [reply samples/deploy/crd/v1beta1/ApisixRoute.yaml]: user can only setting 
```yaml
   authentication:     
       enable: false
```

if set type required, they must set one of `keyAuth` or `basicAuth` type.

Comment (tokers) [reply samples/deploy/crd/v1beta1/ApisixRoute.yaml]: But seems like we don't have a default authentication way. When people enable auth but don't configure the specific authentication, actually no authentication is enabled, this behavior violated the semantic of field `enable`.

Comment (tao12345666333) [reply samples/deploy/crd/v1beta1/ApisixRoute.yaml]: let me set the default value to ｋｅｙＡｕｔｈ

Comment (tokers) [reply samples/deploy/crd/v1beta1/ApisixRoute.yaml]: Got it.

**Review by tokers — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]

**Pull request merged**